fvm/programs/programs.go

```diff
@@ -93,38 +93,23 @@ func (p *Programs) HasChanges() bool {
 	return len(p.programs) > 0 || p.cleaned
 }
 
-func (p *Programs) unsafeForceCleanup() {
-	p.cleaned = true
-
-	// Stop using parent's data to prevent
-	// infinite chaining of objects
-	p.parent = nil
-
-	// start with empty storage
-	p.programs = make(map[common.LocationID]*ProgramEntry)
-}
-
-// ForceCleanup is used to force a complete cleanup
-// It exists temporarily to facilitate a temporary measure which can retry
-// a transaction in case checking fails
-// It should be gone when the extra retry is gone
-func (p *Programs) ForceCleanup() {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	p.unsafeForceCleanup()
-}
-
-func (p *Programs) Cleanup(changedContracts []ContractUpdateKey) {
+func (p *Programs) Cleanup(changedContracts []ContractUpdateKey, hasFreezedAccounts bool) {
 	p.lock.Lock()
 	defer p.lock.Unlock()
 
 	// In mature system, we would track dependencies between contracts
 	// and invalidate only affected ones, possibly setting them to
 	// nil so they will override parent's data, but for now
 	// just throw everything away and use a special flag for this
-	if len(changedContracts) > 0 {
-		p.unsafeForceCleanup()
+	if len(changedContracts) > 0 || hasFreezedAccounts {
+		p.cleaned = true
+
+		// Stop using parent's data to prevent
+		// infinite chaining of objects
+		p.parent = nil
+
+		// start with empty storage
+		p.programs = make(map[common.LocationID]*ProgramEntry)
 		return
 	}
 
```

Comment:
btw, why do we need this?  can we just skip setting the program if the location is not a real address?

Reply:
@janezpodhostnik , @m4ksio ^^ thoughts ?

Reply:
Other non real address locations are, Transaction and Script context afaik. So caching them for sure has some value.

Reply:
This had to do with Cadence requirement that once program is set in the cache, it must be available during the execution.
Not sure if that's still a thing - @turbolent might know

Reply:
@SupunS do you know ?

Reply:
I'm not fully familiar with the program cache implementation tbh, but yeah, cadence expects the FVM to return the same program that was previously set by the cadence. There's a good explanation for this 'informal contract' here:
https://github.com/onflow/cadence/blob/443ea529bed196b0ce0dc138246ee000ad4250fe/runtime/interface.go#L37-L62

Not having this may cause non-determinism. e.g: https://github.com/onflow/cadence/pull/1642/

Reply:
can these entries be imported?  if not, then they are just top level "main" functions, and i'm pretty sure it's safe to not cache the "main"

Reply:
a 'program' in this case is always a contract (an import)

Reply:
can these contracts programs be reused?  if not, we're basically stuck with sequential parsing (or we need a per transaction "cache" for these)

Reply:
yea, program cache is shared across transactions (and blocks?). Say `Tx1` imports contract `A` for the first time, then it will load source code from storage, parse and type check, and caches it in the programs cache. If another transaction `Tx2` imports the same contract `A`, then it will be loaded from the program cache.

Reply:
there may be some miscommunication here.  I understand what the program cache is suppose to do in general.  My question is why do we need to explicitly evict these contract programs after each transaction?  Do their ast program / elaboration change across transactions?

Reply:
as far as I know, script / tx programs are `Set` but never `Get` ( not sure why they are set )

Reply:
it's used here https://github.com/onflow/cadence/blob/master/runtime/runtime.go#L1241.  if we never get from cache, we wouldn't need this fix =P

Reply:
I meant non-contract programs :) Transaction and Script

Reply:
ah right! So I checked the code again, and it seems the currently executing script/transaction is also stored in the cache, not only the imported contracts - which I feel is not necessary. Not sure if it was done intentionally, but looks like an oversight to me.

To me, it sounds like we wouldn't need to store the transaction/script in the first place (should be handled at the cadence side), and could simply remove this check/loop from FVM.

Reply:
@SupunS only possibility here is caching scripts, but then FVM is evicting them. Maybe initial idea was some cache with TTL.

Reply:
ah yes, good point! There were some talk/ideas about re-using scripts. i.e: Sending the changes as script arguments, instead of sending a new script with changes. Perhaps that was the idea. (Sorry I don't know the details)